### PR TITLE
adding AWTToPS

### PR DESCRIPTION
**Pull request by hfazai**

### CODE REVIEW

File: galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt [hunk 1]
```diff
@@ -15,8 +15,1261 @@
...
+  /**
+   * Sets the current color to the specified color. All subsequent graphics operations
+   * will use this specified color.
+   * @param c the color to be set
+   * @see Color
+   *
+   * @see getColor
+   */
+  override fun setColor(c: Color?) {
+    if (c != null) {
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt:141]: c?.let{

File: galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt [hunk 2]
```diff
@@ -15,8 +15,1261 @@
...
+    emitScale(sx.toDouble(), sy.toDouble())
+  }
+
+  /**
+   * Gets the current color.
+   * @see setColor
+   */
+  override fun getColor(): Color {
+    return clr
+  }
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt:118]: turn to single exp fun

File: galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt [hunk 3]
```diff
@@ -15,8 +15,1261 @@
...
+    stream.println(if (fill) "eofill" else "stroke")
+  }
+
+  private fun doRoundRect(x: Int, y: Int, width: Int, height: Int, arcWidth: Int, arcHeight: Int, fill: Boolean) {
+    val y = swapCoord(y)
+
+    emitMoveto(x + arcHeight, y)
+
+    // top, left to right
+    stream.print(x + width)
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt:863]: with(stream){}

File: galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt [hunk 4]
```diff
@@ -15,8 +15,1261 @@
...
+    // translate the page to be centered there
+    emitTranslate(x + width.toFloat() / 2.0, y - height.toFloat() / 2.0)
+    emitScale(1.0, (height.toFloat() / width.toFloat()).toDouble())
+    if (fill) {
+      emitMoveto(0, 0)
+    }
+
+    // now draw the arc.
+    val endAngle = startAngle + arcAngle.toFloat()
+    stream.print("0 0 ")
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt:930]: Add empty line between 
use when(stream){}

File: galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt [hunk 5]
```diff
@@ -15,8 +15,1261 @@
...
+      // nothing
+    } finally {
+      if (!result) {
+        stream.println("%warning: error on image grab")
+        System.err.println("warning: error on image grab: " + pg.status)
+        return false
+      }
+    }
+    gsave()
+    stream.println("currentpoint translate")
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt:977]: use with(stream)

File: galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt [hunk 6]
```diff
@@ -15,8 +15,1261 @@
...
+    emitTranslate(x, y)
+
+    // compute image size. First of all, if width or height is 0, image is 1:1.
+    if (height == 0 || width == 0) {
+      height = pc.dimensions!!.height
+      width = pc.dimensions!!.width
+    }
+    emitScale(width, height)
+    stream.print(pc.dimensions!!.width)
+    stream.print(" ")
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt:1054]: with(stream) ( change it in the rest of the file )

**Review by ZMarouani — CHANGES_REQUESTED** [review]

File: galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt [hunk 7]
```diff
@@ -15,8 +15,1261 @@
...
+   * @param y the y coordinate
+   * @param width the width of the area
+   * @param height the height of the area
+   * @see translate
+   */
+  override fun create(x: Int, y: Int, width: Int, height: Int): Graphics {
+    val graphics = create()
+    graphics.translate(x, y)
+    graphics.clipRect(0, 0, width, height)
+    return graphics
```
[the remark this answers is not in the document]
Comment (hfazai) [reply galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt:86]: For better format , I suggest:
`    val graphics = create()
    return graphics.apply {
      translate(x, y)
      clipRect(0, 0, width, height)
    }`

File: galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt [hunk 8]
```diff
@@ -15,8 +15,1261 @@
...
+   * Translates the specified parameters into the origin of
+   * the graphics context. All subsequent
+   * operations on this graphics context will be relative to this origin.
+   * @param x the x coordinate
+   * @param y the y coordinate
+   * @see scale
+   */
+  override fun translate(x: Int, y: Int) {
+    emitTranslate(x, -y)
+  }
```
[the remark this answers is not in the document]
Comment (hfazai) [reply galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt:99]: it doesnt have a return type

File: galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt [hunk 9]
```diff
@@ -15,8 +15,1261 @@
...
+   * @param y the y coordinate
+   * @param width the width of the rectangle
+   * @param height the height of the rectangle
+   * @see fillRect
+   *
+   * @see clearRect
+   */
+  override fun drawRect(x: Int, y: Int, width: Int, height: Int) {
+    doRect(x, y, width, height, false)
+  }
```
[the remark this answers is not in the document]
Comment (hfazai) [reply galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt:336]: dooesnt have return type

File: galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt [hunk 10]
```diff
@@ -15,8 +15,1261 @@
...
+  /**
+   * Disposes of this graphics context once it is no longer referenced.
+   *
+   * @see dispose
+   */
+  override fun finalize() {
+    dispose()
+  }
+
+  override fun toString(): String = javaClass.name + "[font=" + getFont() + ",color=" + color + "]"
```
[the remark this answers is not in the document]
Comment (hfazai) [reply galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt:712]: We only need the class name. So it's ok

File: galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt (hunk 1 above)
Comment (hfazai) [reply galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt:141]: why using `let`? here we dont want to return something

Comment (ZMarouani) [reply galite-core/src/main/kotlin/org/kopi/galite/util/AWTToPS.kt:140]: using "let" doesn't require to return something ( lambda's return can be void ) and in this case you are free to keep the if ( != null ) too

**Review by ZMarouani — APPROVED** [review]

**Pull request merged**